c++/src/kj/common-test.c++

```diff
@@ -793,29 +794,163 @@ TEST(Common, ArrayAsBytes) {
   }
 }
 
-KJ_TEST("ArrayPtr operator ==") {
-  KJ_EXPECT(ArrayPtr<const int>({123, 456}) == ArrayPtr<const int>({123, 456}));
-  KJ_EXPECT(!(ArrayPtr<const int>({123, 456}) != ArrayPtr<const int>({123, 456})));
-  KJ_EXPECT(ArrayPtr<const int>({123, 456}) != ArrayPtr<const int>({123, 321}));
-  KJ_EXPECT(ArrayPtr<const int>({123, 456}) != ArrayPtr<const int>({123}));
-
-  KJ_EXPECT(ArrayPtr<const int>({123, 456}) == ArrayPtr<const short>({123, 456}));
-  KJ_EXPECT(!(ArrayPtr<const int>({123, 456}) != ArrayPtr<const short>({123, 456})));
-  KJ_EXPECT(ArrayPtr<const int>({123, 456}) != ArrayPtr<const short>({123, 321}));
-  KJ_EXPECT(ArrayPtr<const int>({123, 456}) != ArrayPtr<const short>({123}));
-
-  KJ_EXPECT((ArrayPtr<const StringPtr>({"foo", "bar"}) ==
-             ArrayPtr<const char* const>({"foo", "bar"})));
-  KJ_EXPECT(!(ArrayPtr<const StringPtr>({"foo", "bar"}) !=
-             ArrayPtr<const char* const>({"foo", "bar"})));
-  KJ_EXPECT((ArrayPtr<const StringPtr>({"foo", "bar"}) !=
-             ArrayPtr<const char* const>({"foo", "baz"})));
-  KJ_EXPECT((ArrayPtr<const StringPtr>({"foo", "bar"}) !=
-             ArrayPtr<const char* const>({"foo"})));
-
-  // operator== should not use memcmp for double elements.
-  double d[1] = { nan() };
-  KJ_EXPECT(ArrayPtr<double>(d, 1) != ArrayPtr<double>(d, 1));
+constexpr auto strong_equal = std::strong_ordering::equal;
+constexpr auto strong_less = std::strong_ordering::less;
+constexpr auto strong_greater = std::strong_ordering::greater;
+constexpr auto unordered = std::partial_ordering::unordered;
+constexpr auto partial_equal = std::partial_ordering::equivalent;
+constexpr auto partial_less = std::partial_ordering::less;
+constexpr auto partial_greater = std::partial_ordering::greater;
+
+template<typename A, typename B>
```

Comment:
only one typename needed, right? (here and below)

Reply:
No multiple are needed to do equality comparisons.

Reply:
Why? Because of nullptr? Isn't nullptr convertible to arrayptr? I also wonder if we really need comparison to nullptr rather than another ArrayPtr?

Reply:
no because of things like const StringPtr vs const char* const.